libraries/standard/mqtt/src/mqtt.c

```diff
@@ -99,6 +101,509 @@ static int32_t sendPacket( MQTTContext_t * pContext,
+                                bytesReceived );
+        }
+        else
+        {
+            IotLogErrorWithArgs( "Packet reception failed. ReceivedBytes=%d, "
+                                 "ExpectedBytes=%u.",
+                                 bytesReceived,
+                                 bytesToReceive );
+            status = MQTTRecvFailed;
+        }
+        
+    }
+
+    return status;
+}
+
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief Send acks for received QoS 1/2 publishes.
+ *
+ * @param[in] pContext MQTT Connection context.
+ * @param[in] packetId packet ID of original PUBLISH.
+ * @param[in,out] pPublishState Current/updated publish state in record.
+ *
+ * @return MQTTSuccess or MQTTIllegalState.
+ */
+static MQTTStatus_t sendPublishAcks( MQTTContext_t * const pContext,
+                                     uint16_t packetId,
+                                     MQTTPublishState_t * pPublishState )
+{
+    MQTTStatus_t status = MQTTSuccess;
+    MQTTPublishState_t newState = MQTTStateNull;
+    int32_t bytesSent = 0;
+    uint8_t packetTypeByte = 0U;
+    MQTTPubAckType_t packetType;
+
+    assert( pContext != NULL );
+    assert( pPublishState != NULL );
+
+    switch( *pPublishState )
+    {
+        case MQTTPubAckSend:
+            packetTypeByte = MQTT_PACKET_TYPE_PUBACK;
+            packetType = MQTTPuback;
+            break;
+        case MQTTPubRecSend:
+            packetTypeByte = MQTT_PACKET_TYPE_PUBREC;
+            packetType = MQTTPubrec;
+            break;
+        case MQTTPubRelSend:
+            packetTypeByte = MQTT_PACKET_TYPE_PUBREL;
+            packetType = MQTTPubrel;
+            break;
+        case MQTTPubCompSend:
+            packetTypeByte = MQTT_PACKET_TYPE_PUBCOMP;
+            packetType = MQTTPubcomp;
+            break;
+        default:
+            /* Nothing to send. */
+            break;
+    }
+
+    if( packetTypeByte != 0U )
+    {
+        status = MQTT_SerializeAck( &( pContext->networkBuffer ),
+                                    packetTypeByte,
+                                    packetId );
+        if( status == MQTTSuccess )
+        {
+            bytesSent = sendPacket( pContext,
+                                    pContext->networkBuffer.pBuffer,
+                                    MQTT_PUBLISH_ACK_PACKET_SIZE );
+        }
+
+        if( bytesSent == ( int32_t ) MQTT_PUBLISH_ACK_PACKET_SIZE )
+        {
+            pContext->controlPacketSent = true;
+            newState = MQTT_UpdateStateAck( pContext,
+                                            packetId,
+                                            packetType,
+                                            MQTT_SEND );
+
+            if( newState == MQTTStateNull )
+            {
+                status = MQTTIllegalState;
+            }
+        }
+        else
+        {
+            IotLogErrorWithArgs( "Failed to send ACK packet: SentBytes=%d, "
```

Comment:
Logging which type of Publish packet has failed here would help with debugging problems.

Reply:
Added a log for the packet type byte.